src/LuaTransformer.ts

```diff
@@ -0,0 +1,3289 @@
+import * as path from "path";
+import * as ts from "typescript";
+
+import {CompilerOptions, LuaLibImportKind, LuaTarget} from "./CompilerOptions";
+import {DecoratorKind} from "./Decorator";
+import * as tstl from "./LuaAST";
+import {LuaLib, LuaLibFeature} from "./LuaLib";
+import {ContextType, TSHelper as tsHelper} from "./TSHelper";
+import {TSTLErrors} from "./TSTLErrors";
+
+export type StatementVisitResult = tstl.Statement | tstl.Statement[] | undefined;
+export type ExpressionVisitResult = tstl.Expression | undefined;
+
+export enum ScopeType {
+    Function,
+    Switch,
+    Loop,
+}
+
+interface Scope {
+    type: ScopeType;
+    id: number;
+}
+
+export class LuaTransformer {
+    public luaKeywords: Set<string> = new Set([
+        "and", "break", "do",  "else", "elseif", "end",    "false",  "for",  "function", "if",
+        "in",  "local", "nil", "not",  "or",     "repeat", "return", "then", "until",    "while",
+    ]);
+
+    private selfIdentifier = tstl.createIdentifier("self");
```

Comment:
Don't think this is a good idea actually, because it messes with the s positions and parents. Would be better to rewrite this to a factory method.